packages/ai-jsx/src/react/completion.tsx

```diff
@@ -1,65 +1,73 @@
 import { isJsxBoundary } from './jsx-boundary.js';
+import { memo } from '../core/memoize.js';
 import { AIJSXError, ErrorCode } from '../core/errors.js';
 import z from 'zod';
+import * as CryptoJS from 'crypto-js';
 
-function reactComponentName(component: React.JSXElementConstructor<any> | string) {
+function reactComponentName(component: AI.Component<any> | React.JSXElementConstructor<any> | string) {
   return typeof component === 'string' ? component : component.name;
 }
 
+interface SerializedComponent {
+  tag: string;
+  children: string | (string | SerializedComponent)[];
+  possiblyIncomplete?: boolean;
+}
+
+type ComponentType = AI.Component<any> | React.JSXElementConstructor<any> | string;
+
 export async function* UICompletion(
-  { example, children }: { example: React.ReactNode; children: AI.Node },
+  {
+    reactComponentsDoc,
+    aiComponentsDoc,
+    children,
+  }: {
+    reactComponentsDoc?: React.ReactNode;
+    aiComponentsDoc?: React.ReactNode;
+    children: AI.Node;
+  },
   { render, logger }: AI.ComponentContext
 ) {
   yield '';
-  const reactComponents = new Set<React.JSXElementConstructor<any> | string>();
-  function collectComponents(node: React.ReactNode | AI.Node, inReact: boolean) {
-    if (Array.isArray(node)) {
-      node.forEach((node) => collectComponents(node, inReact));
-    }
 
-    if (React.isValidElement(node)) {
-      if (inReact) {
-        reactComponents.add(node.type);
-      }
+  // TODO: what does the inReact flag do?
+  const aiComponents = collectComponents(aiComponentsDoc, true);
+  const validAIComponentsMap = Object.fromEntries(Array.from(aiComponents).map((c) => [reactComponentName(c), c]));
 
-      const childrenAreReact = (inReact || node.type === AI.React) && !isJsxBoundary(node.type);
-      if ('children' in node.props) {
-        collectComponents(node.props.children, childrenAreReact);
-      }
-    }
+  const reactComponents = collectComponents(reactComponentsDoc, true);
 
-    if (AI.isElement(node)) {
-      const childrenAreReact = (inReact || node.tag === AI.React) && !isJsxBoundary(node.tag);
-      if ('children' in node.props) {
-        collectComponents(node.props.children, childrenAreReact);
-      }
-    }
-  }
+  const allComponents = [...Array.from(reactComponents), ...Array.from(aiComponents)];
+  const validComponentsMap = Object.fromEntries(allComponents.map((c) => [reactComponentName(c), c]));
 
-  collectComponents(example, true);
+  const memoizedComponents = new Map<string, AI.Node>();
 
-  const validComponents = Object.fromEntries(Array.from(reactComponents).map((c) => [reactComponentName(c), c]));
-
-  interface SerializedComponent {
-    name: string;
-    children: (string | SerializedComponent)[];
+  function getOrCreateMemo(serializedComponent: SerializedComponent, nonMemoized: AI.Node): AI.Node {
+    const hash = CryptoJS.SHA256(JSON.stringify(serializedComponent)).toString();
```

Comment:
Why hash the `serializedComponent` JSON, instead of using the JSON as the key directly?

Reply:
Woah, I didn't know I could do that. Actually I am changing this part anyway, since it might `memoize` too much. If you have two images with the same prompt, it's gonna give you back the same image (which is probably not the default behaviour). I'm gonna give it a path-like string as its key instead.